### PR TITLE
Authorization rules

### PR DESCRIPTION
**Pull request by maths29**
Hi, 👋
In this branch, we have :

Installed CanCanCan in my project.
Added a role column to the users' table. Used a migration for this.
A user can delete a post if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.
For this, we implemented the post-deleting functionality. Added the "Delete" button to the view and make sure that only authorized users can see it.
A user can delete a comment if it is theirs or if they have an admin role (column role has value "admin"). Used CanCanCan for this authorization.
For this, we implemented the comment-deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -30,6 +30,20 @@ def index
...
+    authorize! :destroy, @post
+    @author = @post.user
+    @author.decrement!(:posts_counter)
+    @post.destroy
+    if @post.destroy
+      redirect_to user_posts_path(current_user), notice: 'post was successfully deleted.'
+    else
+      redirect_to redirect_url, alert: 'Failed to delete the post.'
+    end
+  end
```
Comment (Ibizugbe) [inline app/controllers/posts_controller.rb:49]: Nice job so far😉. An error is being rendered when I try to delete a post😉.
image

Apart from this, Kindly add the link for deleting a comment😉. This is required to test the delete comment action for admins and users😉. I also noticed that the sign_out button has been removed and I had to manually sign out, for a better user experience and to make it easier for others to test your project, I suggest you add the sign_out button👍🏼.


Lastly, for post-authorization, kindly do well to add ` load_and_authorize_resource` to your post controller😉.

**Review by Ibizugbe — CHANGES_REQUESTED** [review]
Hi @maths29  @Innocent259 👋🏼👋🏼👋🏼, 

request change


Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights✅✅✅

- [x] Linters are passing👍🏼.
- [x] It is possible to assign an admin role to a user👍🏼.
- [x] `cancancan` is present in the gemfile👍🏼.


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -30,6 +32,17 @@ def index
...
+    @post = Post.find(params[:id])
+    authorize! :destroy, @post
+    if @post.destroy
+      flash.now[:success] = 'Post was successfully destroyed!'
+    else
+      flash[:error] = 'Post couldnt be destroyed!'
+    end
+    redirect_to user_posts_path
+  end
+
```
Comment (Munsa1) [inline app/controllers/posts_controller.rb]: - [ ] As requested by the previous reviewer kindly ensure deleting of a post by the user who created it is possible

Videocast
[cinnamon-2023-12-23T153822+0200.webm](https://github.com/maths29/blog_app/assets/81932478/27bad636-1184-46a6-aee5-a01eb1519e49)

- [ ] Also creating a comment now is not possible, kindly fix the error that occurs each time a user  tries to create a comment for a post
[cinnamon-2023-12-23T154014+0200.webm](https://github.com/maths29/blog_app/assets/81932478/e57bded8-1a96-496c-a15c-2c1da5cad832)

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @maths29, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
Passing Linters :heavy_check_mark: 
Correct Git Flow :heavy_check_mark: 
Descriptive ReadMe File :heavy_check_mark: 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Efiamotu-1 — APPROVED** [review]
Hi @maths29 , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# STATUS: APPROVED 🕺🥳 

image

## Highlights :dart: 
* Correct Gitflow :heavy_check_mark: 
* Good commit messages :heavy_check_mark: 
* All linters passing :heavy_check_mark: 
* Good PR summary :heavy_check_mark: 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

## Important Notes 📝

- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

> **Note:** Please, you can now Merge since it is Approved ✅, However if there are any conflicts please resolve them before merging.


## Thank you for your time 🙏 and call me if you need help or clarification ☎️

- This is My **GitHub**: [@Efiamotu-1](https://github.com/Efiamotu-1/)
- This is My **linkedIn**.[Habeeb Efiamotu Musa](https://www.linkedin.com/in/musa-habeeb/)
- This is My **Twitter**: [@EFYAMOTU](https://twitter.com/EFYAMOTU)
- This is My **Slack**: [Habeeb Efiamotu Musa](https://slack.com)

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**